cpp/src/neighbors/scann/detail/scann_quantize.cuh

```diff
@@ -267,6 +269,231 @@ void unpack_codes(raft::resources const& res,
+ * @param noise_shaping_threshold the threshold T in the Theorem
+ * @return eta
+ */
+template <typename IdxT>
+__device__ inline float compute_avq_eta(IdxT dim, const float sq_norm, const float threshold)
+{
+  return (dim - 1) * (threshold * threshold / sq_norm) / (1 - threshold * threshold / sq_norm);
+}
+
+/**
+ * @brief helper to convert a float to bfloat16 (represented as int16_t)
+ *
+ * @param f the float value
+ * @return the bflaot16 value (as int16_t)
+ */
+__device__ inline int16_t float_to_bfloat16(const float& f)
+{
+  nv_bfloat16 val = __float2bfloat16(f);
+  return reinterpret_cast<int16_t&>(val);
+}
+
+/**
+ * @brief helper to convert a bfloat16 (represented as int16_t) to float
+ *
+ * @param bf16 the bf16 value (represented as int16_t)
+ * @return the float value
+ */
+__device__ inline float bfloat16_to_float(int16_t& bf16)
+{
+  nv_bfloat16 nv_bf16 = reinterpret_cast<nv_bfloat16&>(bf16);
+  return __bfloat162float(nv_bf16);
+}
+
+/**
+ * @brief Select the next bfloat16 value to try during coordinate descent
+ *
+ * Based on the signs of the current residual and quantized value,
+ * increment or decrement the quantized value to push residual closer to 0
+ *
+ * @param res the float residual
+ * @param current the current quantized dimension
+ * @return the other possible quantized value
+ */
+__device__ inline int16_t bfloat16_next_delta(float& res, int16_t& current)
+{
+  uint32_t res_sign  = ((int32_t)res & (1u << 31) >> 31);
+  uint32_t curr_sign = (current & (1 << 15)) >> 15;
+
+  if (res_sign == curr_sign) { return current - 1; }
+
+  return current + 1;
+}
+
+/**
+ *
+ */
+template <uint32_t BlockSize, typename IdxT>
+__launch_bounds__(BlockSize) RAFT_KERNEL
+  quantize_bfloat16_noise_shaped_kernel(raft::device_matrix_view<const float, IdxT> dataset,
+                                        raft::device_matrix_view<int16_t, IdxT> bf16_dataset,
+                                        raft::device_vector_view<const float> sq_norms,
+                                        float noise_shaping_threshold)
+{
+  IdxT row_idx = raft::Pow2<32>::div(IdxT{threadIdx.x} + IdxT{BlockSize} * IdxT{blockIdx.x});
+
+  if (row_idx >= dataset.extent(0)) { return; }
+
+  uint32_t lane_id = raft::Pow2<32>::mod(threadIdx.x);
+
+  IdxT dim = dataset.extent(1);
+
+  // 1 / ||x||
+  float inv_norm = 1 / sqrtf(sq_norms[row_idx]);
+  float eta      = compute_avq_eta(dim, sq_norms[row_idx], noise_shaping_threshold);
+
+  // < r, x >
+  float residual_dot = 0.0;
+
+  for (int i = lane_id; i < dim; i += 32) {
+    bf16_dataset(row_idx, i) = float_to_bfloat16(dataset(row_idx, i));
+
+    float residual = dataset(row_idx, i) - bfloat16_to_float(bf16_dataset(row_idx, i));
+    residual_dot += dataset(row_idx, i) * residual * inv_norm;
+  }
+
+  // reduce  and broadcast residual_dot across warp
+  for (uint32_t offset = 16; offset > 0; offset >>= 1) {
+    residual_dot += raft::shfl_xor(residual_dot, offset, 32);
+  }
```

Comment:
Note that we have a [raft::warpReduce](https://github.com/rapidsai/raft/blob/bea333e76bf230bf1e4d94e23ab87ef6e89c1133/cpp/include/raft/util/reduction.cuh#L75) helper

The raft implementation unrolls the loop, I am not sure if that works for you in terms of register usage. Therefore just treat this as an FYI.